src/iotile/cloud.service.ts

```diff
@@ -563,4 +563,94 @@ export class CloudService {
                  return device;
                });
   }
+
+  public fetchDevicesAndVariablesForProject(project: Project): ReplaySubject<any> {
+    let returnedData = new ReplaySubject(1);
+
+    let firstObservable = Observable.forkJoin(
+      this.getVariables(project),
+      this.getDevices(project)
+    );
+
+    firstObservable.subscribe(data => {
+      let variables: Array<Variable> = data[0];
+      let devices: Array<Device> = data[1];
+
+      returnedData.next(project);
+    }, err => {
+      console.error(err);
+      returnedData.error(err);
+    });
+
+    return returnedData;
+  }
+
+  public fetchSensorGraphsForProject(project: Project): ReplaySubject<any> {
+    let returnedData = new ReplaySubject(1);
+    let slugDict = {};
+    let observables = [];
+
+    // 1.- Build a set of unique sensor graphs for all devices in project
+    project && project.devices.forEach(device => {
+      let sgSlug: string = device.sensorGraphSlug;
+      slugDict[sgSlug] = 'found';
+    });
+
+    // 2.- Build a list of required Observables (one per slug)
+    let sgSlugList: Array<string> = Object.keys(slugDict);
+    if (sgSlugList.length === 0) {
+      // Return Project as is if no SGs
+      returnedData.next(project);
+    }
+    console.log('Required SGs: ', sgSlugList);
+    sgSlugList.forEach(slug => {
+      observables.push(this.getSensorGraph(slug));
+    });
+
+    // 3.- Now do the actuall HTTP GETs
+    let firstObservable = Observable.forkJoin(observables);
+    firstObservable.subscribe(data => {
+      let sgMap = {};
+      data.forEach(sg => {
+        let slug: string = sg['slug'];
+        sgMap[slug] = <SensorGraph> sg;
+      });
+
+      // 4.- Finally, update all Device objects with their SensorGraph
+      project.devices.forEach(device => {
+        let sgSlug: string = device.sensorGraphSlug;
+        if (sgMap[sgSlug]) {
+          device.sg = sgMap[sgSlug];
+        } else {
+          console.error('SG ' + sgSlug + ' not found for Device ' + device.slug);
+        }
+      });
+
+      returnedData.next(project);
+    }, err => {
+      console.error(err);
+      returnedData.error(err);
+    });
+
+    return returnedData;
+  }
+
+  public getProject(projectId: string): Observable<any> {
+    let url = '/project/' + projectId;
+    return this._get(url)
+               .map((p: Project) => new Project(p),
+                     err => console.error(err));
+  }
+
+  public fetchProjectWithAssociatedData(projectId): Observable<Project> {
+    let url = '/project/' + projectId;
+
+    return this.getProject(projectId).flatMap((p: Project) => {
+      return this.fetchDevicesAndVariablesForProject(p).flatMap(p => {
```

Comment:
Small suggestion, but you can use a `forkJoin` to get the devicesAndVariables and the sensorGraphs. You need to wait for the project, but once you have the project, you should be able to get the rest together